crates/re_renderer/src/point_cloud_builder.rs

```diff
@@ -141,92 +140,110 @@ impl<'a> PointCloudBatchBuilder<'a> {
-            radius: Size::AUTO,
-        }));
-
-        let num_points = self.0.vertices.len() - old_size;
+        if num_points + self.0.vertices.len() > PointCloudDrawData::MAX_NUM_POINTS {
+            re_log::error_once!(
+                "Reached maximum number of supported points of {}.
+     See also https://github.com/rerun-io/rerun/issues/957",
+                PointCloudDrawData::MAX_NUM_POINTS
+            );
+            num_points = PointCloudDrawData::MAX_NUM_POINTS - self.0.vertices.len();
+        }
+        if num_points == 0 {
+            return self;
+        }
         self.batch_mut().point_count += num_points as u32;
 
-        let new_range = old_size..self.0.vertices.len();
-
-        let max_points = self.0.vertices.len();
-
-        PointsBuilder {
-            vertices: &mut self.0.vertices[new_range],
-            max_points,
-            colors: &mut self.0.color_buffer,
-            picking_instance_ids: &mut self.0.picking_instance_ids_buffer,
-            additional_outline_mask_ids: &mut self
+        {
+            crate::profile_scope!("positions");
+            let num_before = self.0.vertices.len();
+            self.0.vertices.extend(
+                positions
+                    .take(num_points)
+                    .zip(radii.take(num_points))
+                    .map(|(position, radius)| PointCloudVertex { position, radius }),
+            );
+            // Fill up with defaults. Doing this in a separate step is faster than chaining the iterator.
+            let num_default = (self.0.vertices.len() - num_before) - num_points;
+            self.0.vertices.extend(
+                std::iter::repeat(PointCloudVertex {
+                    position: glam::Vec3::ZERO,
+                    radius: Size::AUTO,
+                })
+                .take(num_default),
+            );
+        }
+        {
+            crate::profile_scope!("colors");
+            let num_written = self.0.color_buffer.extend(colors.take(num_points));
+            // Fill up with defaults. Doing this in a separate step is faster than chaining the iterator.
+            self.0
+                .color_buffer
+                .extend(std::iter::repeat(Color32::TRANSPARENT).take(num_written - num_points));
+        }
+        {
+            crate::profile_scope!("picking_instance_ids");
+            let num_written = self
                 .0
-                .batches
-                .last_mut()
-                .unwrap()
-                .additional_outline_mask_ids_vertex_ranges,
-            start_vertex_index: old_size as _,
+                .picking_instance_ids_buffer
+                .extend(picking_instance_ids.take(num_points));
+            // Fill up with defaults. Doing this in a separate step is faster than chaining the iterator.
+            self.0.picking_instance_ids_buffer.extend(
+                std::iter::repeat(PickingLayerInstanceId::default()).take(num_written - num_points),
+            );
```

Comment:
I don't see how these unchecked subtractions could be safe, surely `num_written` can be smaller than `num_points` in many cases?